src/main.js

```diff
@@ -817,7 +839,12 @@ function initializeAfterAppReady() {
       return;
     }
 
-    // get the requesting webContents url
+    // is the requested permission coming from trusted popup window
+    if (isTrustedPopupWindow(webContents)) {
+      callback(true);
```

Comment:
So `/trusted/` is same-origin with the Mattermost server, but I guess it redirects elsewhere and that's why this check is needed? Couldn't you use the [API](https://jitsi.github.io/handbook/docs/dev-guide/dev-guide-iframe) to run the Jitsi webapp same-origin, too, eliminating the need for a change in permissions?

I'm not comfortable granting arbitrary origins arbitrary permissions like this.

Reply:
It is correct, it makes redirect to another domain. API you sent doesn't work for me, even making HTTP proxy on same domain makes too much unecessary network traffice and server load. 

I understand your concern, that is why this is external trusted source. Can we make it to allow only from same domain, not from same origin? Is that something what you would consider?

Reply:
Origins are the only reliable security boundary here. If the API doesn't work, the only alternative I see is explicit origin-specific user approval. That could either be a permission dialog or a JSON config file managed by an org admin. @Willyfrog any thoughts on which direction to go with this?

Reply:
Okay, only security issue is from which origin permissions are requested? You don't have problem if web app communicates with other orgins for wss, https?

Reply:
Communicating with other origins should be fine as long as we don't give those origins any privileged access beyond what they'd gain in a browser.

Reply:
Okay, I will modify this PR. Will remove problematic part of code which you
marked. But I must have popupWindow part. Is that fine with you?

Reply:
I have no security concerns regarding the `popupWindow` code, but I'd still like to better understand why it's needed. I would like to get rid of the `/trusted/` path and make use of existing logic related to plugin paths instead, but that's also not a security concern.

Reply:
Okay, it is not needed if we remove ```parent: mainWindow``` from ```plugins``` popup. If that is fine for you, we can do it that way. With ```parent: mainWindow``` access to audio/video on same origin is broken, I commented out lots of security code in order to figure out what is the issue.

Reply:
> With `parent: mainWindow` access to audio/video on same origin is broken

This sounded very strange to me, so I did some digging. For a minimal setup, I took a release build of the desktop app, started it with the Node debugger (`--inspect`), and ran the following line in the console to patch the permission check:

```JavaScript
require('electron').session.defaultSession.setPermissionRequestHandler((_,__,cb)=>cb(true))
```

Then I added a dummy server that only served a link with `target="_blank"` pointing to a page under `/plugins/` that redirects to `meet.jit.si`. That way I got Jitsi to open in a popup window, hopefully matching your setup on all relevant parts.

Here's what I see: Microphone does not work, as you said. Camera works just fine. Getting rid of `parent` doesn't fix the mic for me, but going into device settings in Jitsi, changing the microphone device, and changing it back again *does work*. After fiddling with the device setting I can enable my mic just fine.

It seems clear to me that this is a bug in Jitsi. It generates a bad constraints object for the `getUserMedia` call until you explicitly specify your hardware. Since it's a bug in Jitsi, I think that's where it should be fixed, too.

Reply:
I appreciate your efforts to debug Electron.
I don't think this is Jitsi bug, as it works in Chrome. Electron uses Chrome.

For the purposes of testing you can load ```meet.jit.si``` in ```contents.loadUrl();``` just to see that it works on Electron, and that bug is not related with Jitsi at all.

Removing ```parent: mainWindow``` did solved device permissions issues for me.

Reply:
> When testing Jitsi important is to have 2 persons in the meeting, to be sure that data is sent and received.

Thanks, that was indeed what I was missing. I see the real issue now. Camera access works locally but video doesn't appear on the other end. Removing `parent` still doesn't fix it for me, though.

Reply:
Which OS are you testing?
I can provide you with pkg for MacOS and deb for Ubuntu. I am on Ubuntu anyway...

Reply:
Thanks for the offer, but I'm on Windows. I have VM setups for Linux and a separate macOS box for testing, but neither of those have cameras.

I've now tried calling `new BrowserWindow` directly from the debugger console, and in that case everything works, including camera, mic, and transmission over the network. The `parent` config still makes no difference.

@Willyfrog any idea what's going on here? I don't want to make a change without understanding it properly.

Reply:
Ah, I think I found the real culprit. The existing `popupWindow` flow sets a custom user agent string, and that messes with the server. If you remove the user agent string it works perfectly. @dpanic can you try doing that with your setup?

@Willyfrog I know the UA string is there to support Google auth and is necessary for the time being. Can you think of a neat way for plugins to bypass/disable the custom UA string?

Reply:
> @Willyfrog any thoughts on which direction to go with this?

I'm a bit divided by this, on one hand having admin setup origins seems like a really useful idea for general use, which could be added either via GPO/buildConfig.js/future-NSUserDefaults, which would provide the best security as if you already trusted them with the platform it would be ok to trust them on what to access. But having the user approve where to go, allows for more flexibility while there might be potential security problems as users sometimes don't read what's going on and the process might be triggered by third parties. So I'm leaning towards the first option.

why not having it hardcoded? because I'd prefer it to be disabled by default only letting those who know what they are doing modify it.

> @Willyfrog any idea what's going on here? I don't want to make a change without understanding it properly.

i have the feeling that jitsi room's and dpanic's ones run a different version of code, as I don't recall getting the constraint error on jitsi rooms.

Reply:
> @Willyfrog I know the UA string is there to support Google auth and is necessary for the time being. Can you think of a neat way for plugins to bypass/disable the custom UA string?

the only place that is really needed right now is on the parent window, as it is used to register against the webapp. Modifying the popup one would only require to ensure google auth keeps working with other plugins, if this wouldn't work we could still manage to have separate popup windows with different user agents as it is already different codepaths (although I'd prefer them to be the same ones)

Reply:
> Ah, I think I found the real culprit. The existing `popupWindow` flow sets a custom user agent string, and that messes with the server. If you remove the user agent string it works perfectly. @dpanic can you try doing that with your setup?
> 
> @Willyfrog I know the UA string is there to support Google auth and is necessary for the time being. Can you think of a neat way for plugins to bypass/disable the custom UA string?

I tested, works now! Great job!
I missed that one, sorry :-)

Reply:
> > @Willyfrog I know the UA string is there to support Google auth and is necessary for the time being. Can you think of a neat way for plugins to bypass/disable the custom UA string?
> 
> the only place that is really needed right now is on the parent window, as it is used to register against the webapp. Modifying the popup one would only require to ensure google auth keeps working with other plugins, if this wouldn't work we could still manage to have separate popup windows with different user agents as it is already different codepaths (although I'd prefer them to be the same ones)

What are next steps?

Do you want me to revert my changes, and modify your current popup window to support with/without user agent override? Is it possible to detect when that is needed?

Two UX questions:
1) Maybe it would be better for popup window to be shown imediatelly even if it's not ready; opinion?
2) Menu bar is better to be hidden, app looks nicer without it; opinion?

Reply:
> Do you want me to revert my changes, and modify your current popup window to support with/without user agent override? Is it possible to detect when that is needed?

what is the needed useragent? do we know what's making it fail? if the resulting UA string is compatible with what plugins need, there wont be a need for having separate logic.

> Maybe it would be better for popup window to be shown imediatelly even if it's not ready; opinion?

I'd advise against, if the window is not ready it might give the wrong impression, is there a need to show it ASAP?

> Menu bar is better to be hidden, app looks nicer without it; opinion?

For the popup? i'm 0/5 about it, but I'd do that on a separate PR to make sure one thing doesn't block the other, it will also allow us to bring UX to take a separate look.

Reply:
> > Do you want me to revert my changes, and modify your current popup window to support with/without user agent override? Is it possible to detect when that is needed?
> 
> what is the needed useragent? do we know what's making it fail? if the resulting UA string is compatible with what plugins need, there wont be a need for having separate logic.
> 
I must check that. But from what I can figure out at the moment, parsing of arbitriary User-Agent could be issue which is breaking it.


> > Maybe it would be better for popup window to be shown imediatelly even if it's not ready; opinion?
> 
> I'd advise against, if the window is not ready it might give the wrong impression, is there a need to show it ASAP?
UX as I mentioned, people click and wait fer seconds for something to happen. That is only reason why I think it is better to show it imediatelly.

> 
> > Menu bar is better to be hidden, app looks nicer without it; opinion?
> 
> For the popup? i'm 0/5 about it, but I'd do that on a separate PR to make sure one thing doesn't block the other, it will also allow us to bring UX to take a separate look.
Okay, in another PR.

Reply:
If we change 

```const nixUA = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome Safari/537.36';```

to
```const nixUA = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/**83.0.4103.97** Safari/537.36';```

Jitsi works, how is this acceptable for you?

Reply:
@dpanic does it work if we use the value from `electron.process.versions.chrome` instead? I'd be wary of hardcoding a version string that doesn't match the underlying framework.

@Willyfrog do you remember if Google OAuth had any browser version requirements?

Reply:
@jupenur they exactly use that to figure out it is not a supported one. Although @larkox might know better if there is a specific version that works or needs to be just `Chrome`

Reply:
If I remember correctly, making the User Agent simply `Chrome` worked at the time so I suppose it does not need any version requirement. Nevertheless, since things might have changed in the meantime, I suppose that it should be tested to make sure.

Reply:
> @dpanic does it work if we use the value from `electron.process.versions.chrome` instead? I'd be wary of hardcoding a version string that doesn't match the underlying framework.
> 
> @Willyfrog do you remember if Google OAuth had any browser version requirements?

```const nixUA = `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/${process.versions.chrome} Safari/537.36`;``` does the job perfectly.

Reply:
I just tested the oauth for zoom with `Chrome/${process.versions.chrome}` and it doesn't work (sorry for spanish, it basically says that this is not secure and that you should try with some other browser)

<img width="783" alt="Screenshot 2020-06-10 at 11 25 25" src="https://user-images.githubusercontent.com/1515906/84251141-3af0b700-ab0d-11ea-9397-04e2abf0b9e6.png">

Reply:
@Willyfrog please check implementation. I changed to what we agreed on.

@jupenur I removed check for isTrustedPopupWindow in setPermissionRequestHandler, please check.